File: Data Structures & Algorithms/jump-game-ii/submission-9.java
class Solution {
    public int jump(int[] nums) {
        int l= 0,r= 0;
        int res = 0;
        while(r<nums.length-1){
            res++;
            int nextR=r;
            for(int i=l;i<=r;i++){
                nextR = Math.max(nextR,nums[i]+i);
            }
            l= r+1;
            r = nextR;
        }
        return res;

        
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-9.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = {}
        if len(nums)<=2:
            return max(nums)
        def dfs(i,N):
            if i in dp:
                return dp[i]
            if i>=N:
                return 0
            dp[i] = max(nums[i]+dfs(i+2,N),dfs(i+1,N))
            return dp[i]
        s = dfs(0,len(nums)-1)
        dp = {}
        return max(s,dfs(1,len(nums)))
        
        